src/utils/resource-container-helper.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../services';
+import {containsNotSelf} from '../dom';
+import {rethrowAsync} from '../log';
+
+/**
+ * Unmount all elements within this container.
+ *
+ * @param {!Element} container
+ */
+export function unmountAllWithin(container) {
+  forAllWithin(container, (element) => element.unmount());
+}
+
+/**
+ * Execute a callback for all elements within the container.
+ *
+ * @param {!Element} container
+ * @param {function(!Element)} callback
+ */
+export function forAllWithin(container, callback) {
+  const resources = Services.resourcesForDoc(container).get();
```

Comment:
@jridgewell @samouri I'd like one advice on this, however. We've used two models on this so far:

1. Shown here: iterate through all elements and check `contains`.
2. In [Owners](https://github.com/ampproject/amphtml/blob/f29f5520d38f7d7df8488e7f3016760a7b894175/src/service/owners-impl.js#L137) and [mutateElement](https://github.com/ampproject/amphtml/blob/f29f5520d38f7d7df8488e7f3016760a7b894175/src/service/mutator-impl.js#L229): call `querySelectorAll`.

A quick set of tradeoffs:
1. all/contains is good when the container has a lot of elements. But if very few, `querySelectorAll` approach might be more efficient.
2. all/contains requires a central place that knows all elements, while `querySelectorAll` uses DOM for this.
3. We have one more use case (see `Owners.scheduleLoad`) where we only do shallow `querySelectorAll`. In those cases we don't want to load deep into the subchildren of an element, instead we only want to load parents and rely on parents to do the right thing. This may eventually become completely unnecessary, but still needed at this time. We can only implement this via `querySelectorAll` since we rely on correct query order.

WDYT?

Reply:
Ok. Agree. I redid this module in the https://github.com/ampproject/amphtml/pull/33450 since that required a few more clarifications. Let's review it in the scope of that pull request.

Reply:
The #33450 has been merged. This file is no longer in this PR.